crypto-serde/src/slice.rs

```diff
@@ -0,0 +1,140 @@
+//! TODO
+//! crypto-serde implementation for slices.
+
+use serde::{Serialize, Serializer};
+
+#[cfg(feature = "alloc")]
+use ::{
+    alloc::vec::Vec,
+    serde::de::Error,
+    serde::{Deserialize, Deserializer},
+};
+
+#[cfg(feature = "zeroize")]
+use zeroize::Zeroize;
+
+/// Serialize the given type as lower case hex when using human-readable
+/// formats or binary if the format is binary.
+pub fn serialize_hex_lower_or_bin<S, T>(value: &T, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+    T: AsRef<[u8]>,
+{
+    if serializer.is_human_readable() {
+        crate::serialize_hex::<_, _, false>(value, serializer)
+    } else {
+        value.as_ref().serialize(serializer)
+    }
+}
+
+/// Serialize the given type as upper case hex when using human-readable
+/// formats or binary if the format is binary.
+pub fn serialize_hex_upper_or_bin<S, T>(value: &T, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+    T: AsRef<[u8]>,
+{
+    if serializer.is_human_readable() {
+        crate::serialize_hex::<_, _, true>(value, serializer)
+    } else {
+        value.as_ref().serialize(serializer)
+    }
+}
+
+/// Deserialize the given slice from hex when using human-readable formats or
+/// binary if the format is binary.
+#[cfg(feature = "alloc")]
+#[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+pub fn deserialize_hex_or_bin<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    if deserializer.is_human_readable() {
+        base16ct::mixed::decode_vec(<&str>::deserialize(deserializer)?).map_err(D::Error::custom)
```

Comment:
`<&str>::deserialize(deserializer)` this will blow up if a serialization format does something funny with strings and can't just borrow the bytes.

Explained in reverse: I believe this only works for serializations formats that serialize strings into bytes as is without modification.

To fix this, we would have to implement a `Visitor`.

Reply:
I think this is a “cross that bridge when we get there” situation. While it’s a hypothetical, in absence of a major format implementation that actually does that I don’t think it’s worth concerning ourselves with.

Reply:
Sounds good to me.

Reply:
Some more thoughts I had here:
- This can also fail with `DeserializeOwned`.
- To a degree this might be useful to make sure that the serialization format doesn't actually do something with the string serialization, as that would probably break constant-time assurances and circumvent the whole thing we are doing here.

Reply:
> This can also fail with DeserializeOwned.

There's nothing magic about `DeserializeOwned`. It's just a marker trait w\ HRTB + a blanket impl which ensures that the `Deserialize` impl is valid for all lifetimes.

This means it's perfectly fine to *temporarily* borrow data within the deserializer, so long as it's only borrowed for the lifetime of the deserializer function and not referenced in the resulting value.

> To a degree this might be useful to make sure that the serialization format doesn't actually do something with the string serialization, as that would probably break constant-time assurances and circumvent the whole thing we are doing here.

I'm not sure what we can really do there. Once the string has been parsed by the input format during deserialization, or handed off to the serializer, it's really out of our hands. The most `crypto-serde` can do is minimize harm.

Reply:
> This means it's perfectly fine to _temporarily_ borrow data within the deserializer, so long as it's only borrowed for the lifetime of the deserializer function and not referenced in the resulting value.

This isn't the experience I had, see https://github.com/RustCrypto/crypto-bigint/pull/73#issuecomment-1079139655. But maybe this is a "bug" in bincode?

Reply:
Yes, that has nothing to do with `DeserializeOwned`. As I said on that issue, `bincode` doesn't support borrowing a slice from the input.

Empirically the human readable formats we're currently testing against (`serde_json` and `toml`) do support this, so until we encounter a format that doesn't it's not an issue.

Reply:
> Yes, that has nothing to do with `DeserializeOwned`. As I said on that issue, `bincode` doesn't support borrowing a slice from the input.

When you don't use `deserialize_from` but `deserialize` it works, I assumed it was because of `DeserializeOwned`. I didn't dig into it, so it being `DeserializeOwned` was just a guess.

> Empirically the human readable formats we're currently testing against (`serde_json` and `toml`) do support this, so until we encounter a format that doesn't it's not an issue.

Yeah I wasn't arguing for that, I just wanted to clear up left-over threads of conversation :smiley:.